ibc/client/submodule.go

```diff
@@ -0,0 +1,225 @@
+
+	for _, option := range options {
+		option(c)
+	}
+
+	c.logger.Info().Msg("👨 Creating Client Manager 👨")
+
+	bus.RegisterModule(c)
+
+	return c, nil
+}
+
+func (c *clientManager) GetModuleName() string { return modules.ClientManagerModuleName }
+
+// CreateClient creates a new client with the given client state and initial
+// consensus state and initialises it with a unique identifier in the IBC client
+// store and emits an event to the Event Logger
+func (c *clientManager) CreateClient(
+	clientState modules.ClientState, consensusState modules.ConsensusState,
+) (string, error) {
+	// Check if the client type is allowed
+	if !isAllowedClientType(clientState.ClientType()) {
+		return "", fmt.Errorf("client type %s is not supported", clientState.ClientType())
+	}
+
+	// Generate a unique identifier for the client
+	identifier := path.GenerateClientIdentifier()
+
+	// Retrieve the client store prefixed with the client identifier
+	prefixed := path.ApplyPrefix(core_types.CommitmentPrefix(path.KeyClientStorePrefix), identifier)
+	clientStore, err := c.GetBus().GetIBCHost().GetProvableStore(string(prefixed))
+	if err != nil {
+		return "", err
+	}
+
+	// Initialise the client with the clientState provided
+	if err := clientState.Initialise(clientStore, consensusState); err != nil {
+		c.logger.Error().Err(err).Str("identifier", identifier).
+			Msg("failed to initialize client")
+		return "", err
+	}
+
+	c.logger.Info().Str("identifier", identifier).
+		Str("height", clientState.GetLatestHeight().ToString()).
+		Msg("client created at height")
+
+	// Emit the create client event to the event logger
+	if err := c.emitCreateClientEvent(identifier, clientState); err != nil {
+		c.logger.Error().Err(err).Str("identifier", identifier).
+			Msg("failed to emit client created event")
+		return "", err
+	}
+
+	return identifier, nil
+}
+
+// UpdateClient updates an existing client with the given identifier using the
+// ClientMessage provided
+func (c *clientManager) UpdateClient(
+	identifier string, clientMessage modules.ClientMessage,
+) error {
+	// Get the client state
+	clientState, err := c.GetClientState(identifier)
+	if err != nil {
+		return err
+	}
+
+	// Retrieve the client store prefixed with the client identifier
+	prefixed := path.ApplyPrefix(core_types.CommitmentPrefix(path.KeyClientStorePrefix), identifier)
+	clientStore, err := c.GetBus().GetIBCHost().GetProvableStore(string(prefixed))
+	if err != nil {
+		return err
+	}
+
+	// Check the state is active
+	if clientState.Status(clientStore) != modules.ActiveStatus {
+		return core_types.ErrIBCClientNotActive()
+	}
+
+	// Verify the client message
+	if err := clientState.VerifyClientMessage(clientStore, clientMessage); err != nil {
+		return err
+	}
+
+	// Check for misbehaviour on the source chain
+	misbehaved := clientState.CheckForMisbehaviour(clientStore, clientMessage)
+	if misbehaved {
+		if err := clientState.UpdateStateOnMisbehaviour(clientStore, clientMessage); err != nil {
```

Comment:
Thoughts on `FreezeClientOnMisbehaviour`?

Not a strong preference, just trying to make things more self explanatory.